src/main/java/uk/ac/ebi/ena/webin/cli/service/SampleService.java

```diff
@@ -139,38 +139,87 @@ protected SampleService(Builder builder) {
+      sraSample = getSraSample(sampleId);
+    } catch (ServiceException e) {
+      LOGGER.error("Failed to retrieve SRA sample: {}", sampleId, e);
 
-      isBiosamplesRetrievalAlreadyAttempted = true;
+      sraSample = null;
     }
 
-    // If the sample couldn't be retrieved from Biosamples above then retrieve it from ENA.
-    Sample sraSample = getSraSample(sampleId);
     if (sraSample == null) {
-      return null;
+      sampleRetrievalMessages.add(ServiceMessage.SAMPLE_SERVICE_VALIDATION_ERROR.format(sampleId));
+      sraSampleRetrievalFailed = true;
     }
 
-    // If SRA sample has a Biosamples accession then retrieve it from Biosamples using this
-    // accession. This is
-    // becuase getting samples data from Biosamples is always preferred.
-    if (sraSample.getBioSampleId() != null && !isBiosamplesRetrievalAlreadyAttempted) {
-      Sample biosamplesSample = getBiosamplesSample(sraSample.getBioSampleId());
-      if (isBiosamplesSampleValid(biosamplesSample)) {
+    // If an SRA sample has a Biosamples accession, then try to retrieve it from Biosamples using
+    // this accession. This is because getting samples data from Biosamples is always preferred.
+    // This is done only when a previous attempt to retrieve from BioSamples is not performed, so
+    // there
+    // is no risk of duplicate checks and duplicate messages getting added to the list
+    if (sraSample != null
+        && sraSample.getBioSampleId() != null
+        && !isBiosamplesRetrievalAlreadyAttempted) {
+      biosamplesSample = getBiosamplesSample(sraSample.getBioSampleId());
+
+      if (biosamplesSample == null) {
+        sampleRetrievalMessages.add(ServiceMessage.BIOSAMPLES_NOT_FOUND_MESSAGE.format(sampleId));
+      } else if (isBiosamplesSampleValid(biosamplesSample)) {
         return biosamplesSample;
+      } else {
+        sampleRetrievalMessages.add(ServiceMessage.BIOSAMPLES_INVALID_MESSAGE.format(sampleId));
       }
     }
 
-    // Getting here means we couldn't get sample from Biosamples. So return the SRA sample instead
-    // after adding
-    // attribute information to it.
-    Sample sampleFromXml = sampleXmlService.getSample(sraSample.getSraSampleId());
-    sraSample.setAttributes(sampleFromXml.getAttributes());
+    // Getting here means we couldn't get a sample from Biosamples. So return the SRA sample instead
+    // after adding attribute information to it. This is performed if SRA sample retrieval is not
+    // attempted earlier
+    if (!sraSampleRetrievalFailed) {
+      Sample sampleFromXml = sampleXmlService.getSample(sraSample.getSraSampleId());
+
+      if (sampleFromXml == null) {
+        sampleRetrievalMessages.add(
+            ServiceMessage.SAMPLE_SERVICE_VALIDATION_ERROR.format(sampleId));
+
+        throw new ServiceException(sampleRetrievalMessages);
+      }
+
+      sraSample.setAttributes(sampleFromXml.getAttributes());
+    } else {
+      throw new ServiceException(sampleRetrievalMessages);
```

Comment:
best to just return null then throwing exceptions in this method. this will be compatible with webin-cli code that uses this method.

Reply:
Agreed.